src/java/htsjdk/samtools/SAMRecord.java

```diff
@@ -382,20 +376,14 @@ else if (null != mHeader) {
      * @return Index in the sequence dictionary of the reference sequence. If the read has no reference sequence, or if
      * the reference name is not found in the sequence index, NO_ALIGNMENT_REFERENCE_INDEX (-1) is returned.
      *
-     * @throws IllegalStateException if the reference index cannot be resolved because the SAMFileHeader for the
-     * record is null.
+     * @throws IllegalStateException if the reference index must be resolved but cannot be because the SAMFileHeader
+     * for the record is null.
      */
     public Integer getReferenceIndex() {
-        if (null == mReferenceIndex) {
-            // try to resolve the reference index
-            if (NO_ALIGNMENT_REFERENCE_NAME.equals(mReferenceName)) {
-                mReferenceIndex = NO_ALIGNMENT_REFERENCE_INDEX;
-            }
-            else if (null != mHeader) {
-                mReferenceIndex = mHeader.getSequenceIndex(mReferenceName);
-            }
-            else {
-                throw new IllegalStateException("A non-null SAMFileHeader is required to resolve the reference index");
+        if (null == mReferenceIndex) { // try to resolve the reference index
+            mReferenceIndex = resolveIndexFromName(mReferenceName, mHeader, false);
+            if (null == mReferenceIndex) {
+                return NO_ALIGNMENT_REFERENCE_INDEX;
```

Comment:
This new code leaves `mReferenceIndex` set to null in the case where the lookup failed, whereas the previous version would leave it set to `NO_ALIGNMENT_REFERENCE_INDEX`. Is this what we want?